### ISSUE TITLE
Clevo Notebooks Support

### ISSUE THREAD

**Issue by rkkoszewski** [antonpup/Aurora#153]
Hi ,

So I implemented some basic support for Clevo devices for Aurora as @antonpup suggested me, but I have some trouble with Aurora recognizing the device. I hope somebody can guide or help me to make it work.

Here is the Pull Request https://github.com/antonpup/Aurora/pull/152

I have 2 issues:

1: The Clevo device is marked as "Initialized" correctly, but Aurora still says "No Keyboard Detected". Any idea?

2: Does Aurora support region based lightning instead of per key lightning?

Greetings

**Comment by antonpup** [comment]
Hey! Thanks for contributing!

1. The devices aren't actually tied to the virtual keyboard you see within Aurora. The user can select for example Logitech keyboard to be displayed, but Aurora will update both Logitech and Corsair keyboards (if both are connected at the same time).

2. At the moment, Aurora does not support this. However, it is something that I need to do, it will then also allow LightFX keyboard lighting. Please check out the [dev branch](https://github.com/antonpup/Aurora/tree/dev), I will change the `bool UpdateDevice(Dictionary<DeviceKeys, Color> keyColors, bool forced = false);` to `bool UpdateDevice(DeviceColorComposition keyColors, bool forced = false);` Where DeviceColorComposition will be a struct with Dictionary<DeviceKeys, Color> and a resulting bitmap. You will be able to get an average color of a select bitmap region and use it for lighting.

I will provide some sample code here, once I've committed this change. And please, make pull requests for the dev branch instead of master.

Thank you.

**Comment by antonpup** [comment]
Ok, I pushed the commit with color retrieval for regions. [https://github.com/antonpup/Aurora/commit/29ae9409b3c97b745b02c1cc1682e46b08c3fb1c](https://github.com/antonpup/Aurora/commit/29ae9409b3c97b745b02c1cc1682e46b08c3fb1c)
I hope this helps!

Here's sample usage:
```C#
public bool UpdateDevice(DeviceColorComposition colorComposition, bool forced = false)
    {
      try
      {
        //Example region - From ESC to Middle of spacebar
        BitmapRectangle keymap_esc = Effects.GetBitmappingFromDeviceKey(DeviceKeys.ESC);
        BitmapRectangle keymap_space = Effects.GetBitmappingFromDeviceKey(DeviceKeys.SPACE);
        PointF spacebar_center = keymap_space.Center;

        BitmapRectangle keymap_esc_to_space = new BitmapRectangle(keymap_esc.Left, keymap_esc.Top, (int)spacebar_center.X - keymap_esc.Left, (int)spacebar_center.Y - keymap_esc.Top);

        Color region_esc_space_color = Utils.BitmapUtils.GetRegionColor(colorComposition.keyBitmap, keymap_esc_to_space);

        //Use the color with your device
        //DeviceSDK.SetRegionColor(region_esc_space_color);
        //etc...

        Global.logger.LogLine($"region_esc_space_color = {region_esc_space_color}");
      }
      catch(Exception exc)
      {
        Global.logger.LogLine("Error when updating device. Error: " + exc, Logging_Level.Error);
        crashed = true;
        isInitialized = false;

        return false;
      }

      return true;
    }
```

**Comment by antonpup** [comment]
Forgot to mention. Feel free to alter other code in Aurora if you need more Utils or you have improvements.

**Comment by rkkoszewski** [comment]
Hi.

Thanks a lot @antonpup for the support. It's a real joy to work like this.

Here is the latest implementation merged into DEV: https://github.com/antonpup/Aurora/pull/155

I have still two issues there:

1: The main issue is with the right region. I'm using a region between F11 and NUMPAD ENTER, but the average color is very dark (almost black). Any idea what could be causing it? 

2: I don't know if this is a bug or it's still being implemented, but I'm using DeviceKeys.Peripheral for the touchpad colors. This was working pretty nicely from the MASTER branch, but in DEV it seems that it's not being used anymore and the device color is always black.

I test it using the GTA Police Effect. In MASTER it made the Touchpad go Red/Blue, but in DEV its just black.

Using DeviceKeys.Peripheral_Logo works, but as the mouse is assumed to be on the right side, it looks weird when the touchpad is right under the keyboard. What would be the best approach here?

**Comment by antonpup** [comment]
I was mid replying to your pr, so I will address all 3 points here.

1. I think the problem for a "darker color" is due to the fact that the region has a lot of empty space, thus more black color. What you could do is either get color from F11 to Right arrow and get color from Numlock to Numpad Enter. And blend the two colors. using `Color color_blend = Utils.ColorUtils.BlendColors(_color1, _color2, 0.5);` Another method would involve retrieving color from 4 regions, for more color accuracy. So F11 to Ctrl, Print Screen to Page Down, Numpad Area, and the arrow keys. That way you would be getting the least amount of black color.
2. The dev branch is undergoing Peripheral device changes. I am adding mouse layouts instead of the generic "Mouse/Headset" button. As a result, peripheral devices (specifically mice) were split into Logo, Scroll Wheel, Frontlight, etc. Long story short, I will adjust the DeviceKeys.Peripheral to apply a single color to all peripheral lights, like before.
3. You don't have to restore back the colors. Unless the device does not revert the colors back after sdk release.

As for the mousepad location. On dev branch, I introduced a way to add different components to the virtual keyboard layout. So for example, in settings you can select a mouse and that mouse is actually added onto the keyboard. Same thing with additional keyboard features like G-keys, those are added on to a standard keyboard layout. So you can now add a "touchpad image" and define a lighting region with it using a mask. The problem is that you mentioned that the touchpad is below the keyboard, which would mean that the virtual keyboard would require more height. And there's not much room to spare for height. So.. we can try and add a touchpad below the keyboard and see how it looks, depending on how big it is, there will be a tiny keyboard. I can set up the touchpad and keyboard modules for you, if you could provide me with a sample picture (you don't have to take one, a link to a store page will be fine) from a top down view, so I can adjust it in Photoshop.

**Comment by rkkoszewski** [comment]
@antonpup I would suggest to show a wide button which would look like "Space" under the keyboard with a label stating that it's the "Touch Pad". That way it wouldn't take up so much space.

**Comment by antonpup** [comment]
@rkkoszewski Right below the spacebar? Same width and height as the spacebar?

![](http://www.xoticpc.com/media/catalog/product/cache/1/thumbnail/9df78eab33525d08d6e5fb8d27136e95/n/e/new_sager_8173_photo_06_1.png)
![](http://www.xoticpc.com/media/catalog/product/cache/1/thumbnail/9df78eab33525d08d6e5fb8d27136e95/n/e/new_sager_9152_photo_06.png)
Also, I am looking at the Clevo Notebooks, and their keyboard layout is very compact. I will see if I can replicate it with Aurora, that way you won't have issues with darker colors.

**Comment by rkkoszewski** [comment]
Yes, mine has the same compact layout.

The touchpad is similar in size as the spacebar, although my touchpad is exactly in the center (horizontally) of the notebook but that shouldn't make a big difference.

It would be really awesome to have a Clevo like layout. 


To answer point 3 (revert keyboard colors to previous ones). Clevo doesn't provide any official SDK, so in order to change the KB colors I talk directly to the Clevo Driver using WMI (That's why I need to run Aurora with Admin privileges so it would work). But the Clevo Hotkey software saves the colors that are set with the official Clevo software to the registry and is easily readable.

**Comment by antonpup** [comment]
I am not sure if I will be able to manipulate some keyboard layouts to fit this compact size. As I need to make shift and space bar smaller to make room for the arrow keys. For example, the Japanese layout doesn't have any room for size changes at all! So making a "compact version" will not be possible with it.
![Japanese layout](http://puu.sh/rmMEv/f1a10b66d2.png)

I will, however, add the touchpad in a few moments.

**Comment by antonpup** [comment]
Had some trouble, I pushed a commit with the touchpad and a peripheral light fix on the dev branch.

**Comment by rkkoszewski** [comment]
Hi @antonpup, thanks! It just works perfectly now!! The police effects in GTA V or the bomb effect looks plain awesome on the touchpad.

Here is my latest update: https://github.com/antonpup/Aurora/pull/155

I have improved the right region of the keyboard (still needs some adjustments, just don't have time right now, maybe tonight I could finish it).

I have finally implemented the "Shutdown" function, so the keyboard now restores its original colors when Aurora is exited.

I would say that the Clevo implementation is 90-95% done. 

Stuff left to do:

- Final adjusts to the right area colors.
- Write a manual page or something stating that you have to run Aurora with Admin Privileges if you want to use the Clevo Support. (Eventually, there is a way to change the WMI permissions, but as of now I have no idea how to do it, nor if it is even possible to change programatically. I will investigate on this later when I have some more free time), and that you need the latest Clevo EC Firmware (which is required to run Windows 10 anyway if I'm not wrong). 
- Do some testing with other Clevo Notebook users and check if its working on other models.

**Comment by antonpup** [comment]
1. You can do it! I believe in you! (If it turns out to be inefficient, I will look into optimizing it)
2. I do have "first-time instructions", which I need to revise. But it can state there that Aurora needs to run as admin. Additionally, Clevo device can check if it has elevated privileges and provide a warning message (or write to log) that Clevo might not work properly without admin rights.
3. That I cannot do, sorry. This is the first time I am hearing about this laptop. >.< So I don't have anyone who has these laptops. But I do plan on having a testing period before releasing next major version, as a lot has changed, so I want to get some fine-tuned feedback from users before officially tossing it out into the public.

**Comment by rkkoszewski** [comment]
So I finally managed to get the right side of the keyboard working correctly. I thought that BitmapRectangle was using start coordinates and end coordinates instead of an initial coordinate, with and height. So now it's working properly.

https://github.com/rkkoszewski/Aurora/commit/5770f3fdacff9ee336e7f0dad68b3609a8222983
